api/src/main/java/org/pmiops/workbench/db/dao/DataSetServiceImpl.java

```diff
@@ -664,14 +664,58 @@ public static String capitalizeFirstCharacterOnly(String text) {
     return StringUtils.capitalize(text.toLowerCase());
   }
 
+  private static String generateSqlWithEnvironmentVariables(
+      String query, KernelTypeEnum kernelTypeEnum) {
+    switch (kernelTypeEnum) {
+      case PYTHON:
+        return query.replaceAll(
+            "\\$\\{projectId}.\\$\\{dataSetId}", "\"\"\" + os.environ[\"WORKSPACE_CDR\"] + \"\"\"");
+      case R:
+        return query.replaceAll(
+            "\\$\\{projectId}.\\$\\{dataSetId}", "\", Sys.getenv(\"WORKSPACE_CDR\"), \"");
+      default:
+        return query;
+    }
+  }
+
+  private static String fillInQueryParams(
+      String query, Map<String, QueryParameterValue> queryParameterValueMap) {
+    StringBuilder stringBuilder = new StringBuilder(query);
+    queryParameterValueMap.forEach(
+        (key, value) -> {
+          if (StandardSQLTypeName.ARRAY.equals(value.getType())) {
```

Comment:
I don't love inlining these into the SQL itself. I think string substitution within the notebook code would be more reusable / understandable / readable. In R: str_glue, in Py `f""`

Alternatively, the parameterized queries in Python were perhaps an ideal way to show how to properly use BigQuery, but understood that diverging this much between R and Python could be a problem.

Reply:
I am happy using str_glue or `f""`, but honestly, I'm unsure what benefit it gives over the config or query parameters (They're both variables that get substitutions in)

I can revert back to that if we like. BigRQuery does just have it use a parameter option instead of a configuration option, so it does trim a lot of boilerplate and just make it basically a substitution.

Reply:
The benefit is that it can work with arrays in R, otherwise I'd prefer config/query parameters.